src/lib/communication/DY.Lib.Communication.Core.fst

```diff
@@ -0,0 +1,278 @@
+  | CommAuthSendMsg: sender:principal -> payload:bytes -> communication_event
+  | CommAuthReceiveMsg: sender:principal -> receiver:principal -> payload:bytes -> communication_event
+  | CommConfAuthSendMsg: sender:principal -> receiver:principal -> payload:bytes -> communication_event
+  | CommConfAuthReceiveMsg: sender:principal -> receiver:principal -> payload:bytes -> communication_event
+  
+
+#push-options "--ifuel 1 --fuel 0"
+%splice [ps_communication_event] (gen_parser (`communication_event))
+%splice [ps_communication_event_is_well_formed] (gen_is_well_formed_lemma (`communication_event))
+#pop-options
+
+instance event_communication_event: event communication_event = {
+  tag = "DY.Lib.Communication.Event";
+  format = mk_parseable_serializeable ps_communication_event;
+}
+
+
+(*** Layer Setup ***)
+
+val comm_layer_pkenc_tag: string
+let comm_layer_pkenc_tag = "DY.Lib.Communication.PkEnc.PublicKey"
+
+val comm_layer_sign_tag: string
+let comm_layer_sign_tag = "DY.Lib.Communication.Sign.PublicKey"
+
+type communication_keys_sess_ids = {
+  pki: state_id;
+  private_keys: state_id;
+}
+
+val comm_label: principal -> principal -> label
+let comm_label sender receiver = join (principal_label sender) (principal_label receiver)
+
+
+(*** Communication Layer ***)
+
+(**** Confidential Send and Receive Functions ****)
+
+val encrypt_message: bytes -> bytes -> bytes -> bytes
+let encrypt_message pk_receiver nonce payload =
+  pk_enc pk_receiver nonce payload
+
+val send_confidential:
+  communication_keys_sess_ids ->
+  principal -> principal -> bytes ->
+  traceful (option timestamp)
+let send_confidential comm_keys_ids sender receiver payload =
+  let*? pk_receiver = get_public_key sender comm_keys_ids.pki (LongTermPkEncKey comm_layer_pkenc_tag) receiver in
+  let* nonce = mk_rand PkNonce (long_term_key_label sender) 32 in
+  trigger_event sender (CommConfSendMsg sender receiver payload);*
+  let msg_encrypted = encrypt_message pk_receiver nonce payload in
+  let* msg_id = send_msg msg_encrypted in
+  return (Some msg_id)
+
+
+val decrypt_message: bytes -> bytes -> option bytes
+let decrypt_message sk_receiver msg_encrypted =
+  pk_dec sk_receiver msg_encrypted
+
+val receive_confidential:
+  communication_keys_sess_ids ->
+  principal -> timestamp ->
+  traceful (option bytes)
+let receive_confidential comm_keys_ids receiver msg_id =
+  let*? sk_receiver = get_private_key receiver comm_keys_ids.private_keys (LongTermPkEncKey comm_layer_pkenc_tag) in
+  let*? msg_encrypted = recv_msg msg_id in
+  let*? payload = return (decrypt_message sk_receiver msg_encrypted) in
+  trigger_event receiver (CommConfReceiveMsg receiver payload);*
+  return (Some payload)
+
+
+(**** Authenticated Send and Receive Functions ****)
+
+val sign_message: principal -> principal -> bytes -> bytes -> bool -> bytes -> bytes -> bytes
+let sign_message sender receiver payload pk_receiver is_encrypted sk nonce =
```

Comment:
Rather than taking `pk_receiver` and `is_encrypted` as separate arguments (where `pk_receiver` seems to always be `empty` if `is_encrypted` is `false`), maybe instead take an `option bytes` for `pk_receiver`, and pass in `None` for the unencrypted case?